cmd/openshift-install/gather.go

```diff
@@ -197,3 +202,36 @@ func unSupportedPlatformGather(directory string) error {
 
 	return logGatherBootstrap(gatherBootstrapOpts.bootstrap, 22, gatherBootstrapOpts.masters, directory)
 }
+
+func logClusterOperatorConditions(ctx context.Context, config *rest.Config) error {
+	client, err := configclient.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "creating a config client")
+	}
+
+	operators, err := client.ConfigV1().ClusterOperators().List(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "listing ClusterOperator objects")
+	}
+
+	for _, operator := range operators.Items {
+		for _, condition := range operator.Status.Conditions {
+			if condition.Type == configv1.OperatorUpgradeable {
+				continue
+			} else if condition.Type == configv1.OperatorAvailable && condition.Status == configv1.ConditionTrue {
+				continue
+			} else if (condition.Type == configv1.OperatorDegraded || condition.Type == configv1.OperatorProgressing) && condition.Status == configv1.ConditionFalse {
+				continue
+			}
+			if condition.Type == configv1.OperatorDegraded {
```

Comment:
Switch case if you would.  This doesn't look like it handles Unknown for status or new condition types.

Reply:
> Switch case if you would.   This doesn't look like it handles Unknown for status or new condition types.

How about a generic:

```
logrus.Infof("Cluster operator {} {} is {} with {}: {}", operator.ObjectMeta.Name, condition.Type, condition.Status, condition.Reason, condition.Message)
```

regardless of `condition.Status`?  I'm also fine returning to `WithFields`, like I had before [this discussion][1].

[1]: https://github.com/openshift/installer/pull/2450#discussion_r330867075

Reply:
hrm.  I guess that loses `Errorf` vs. `Infof`...

Reply:
Ok, I've pushed a4e907c8e -> de5a26aa4 with:

```go
if condition.Type == configv1.OperatorDegraded {
  logrus.Errorf("Cluster operator {} {} is {} with {}: {}", operator.ObjectMeta.Name, condition.Type, condition.Status, condition.Reason, condition.Message)
} else {
  logrus.Infof("Cluster operator {} {} is {} with {}: {}", operator.ObjectMeta.Name, condition.Type, condition.Status, condition.Reason, condition.Message)
}
```

So now I no longer attempt to translate `condition.Status`, and am just highlighting surprising degraded values with `Errorf` with everything else on `Infof`.  How's that look, @abhinavdahiya, @deads2k?

Reply:
available==false needs to be error too.

We only need to log ones where 

```go
good := available == true && progressing==false && degraded==false
if !good{
log.errorf()
}
```

Reply:
> available==false needs to be error too.

Not necessarily, could just be pulling over a slow network or some such.  available=false may be why we're timing out, but it's still not inherently an error.  If a particular operator considers it to have been going on for long enough to be an error, they're free to set degraded to that effect.

> We only need to log ones where
>
> ```go
> good := available == true && progressing==false && degraded==false
> if !good{
>   log.errorf()
> }
> ```

I'm [not logging Upgradeable regardless of condition][1], but that would not match your `good` condition.  Do you think I should log it?  Previous discussion [here][2].  Otherwise, [those are the conditions I'm logging][3].  But I don't think they're all errors.  For another info-level condition, see the Disabled condition from [this test][4].  And there's no obvious reason I can see to assume new, unknown to the installer conditions are errors, so I would prefer to fall back to `Infof` and only promote things we know are errors to `Errorf`.  Our default log level is info, so these will all show up in a user's terminal either way.

[1]: https://github.com/openshift/installer/commit/de5a26aa42df631124db3472ed2b6202fdb21836#diff-be4790dcfa83ca14e7942b4c3b50c5f9R219
[2]: https://github.com/openshift/installer/pull/2450#discussion_r330867229
[3]: https://github.com/openshift/installer/commit/de5a26aa42df631124db3472ed2b6202fdb21836#diff-be4790dcfa83ca14e7942b4c3b50c5f9R221-R224
[4]: https://github.com/openshift/installer/pull/2450#issuecomment-537728581

Reply:
Been a quiet few days here.  Can we just land this as it stands and if necessary twiddle with the output format later?  I think what we have here now is a lot better than not logging anything about ClusterOperator conditions, which is where master is now.

Reply:
I feel like available==false should be an error as well. I feel like degraded but progressing should also be logged. This is only happening after waiting 30 minutes for a cluster to have formed, correct?

That said, I'd prefer we get this in sooner rather than later so we can start looking at whether or not this starts producing the results we'd like to see. We can tweak reporting in the future.

Reply:
The if/then/else logic is a little confusing. What precisely is the installer using to decide "not ready"? Is it not possible to have available=true && degraded=true? Maybe move to a switch.
```
switch {
case condition.Type == configv1.OperatorDegraded && condition.Status == configv1.ConditionTrue:
  logrus.Errorf("Cluster operator {} {} is {} with {}: {}", operator.ObjectMeta.Name, condition.Type, condition.Status, condition.Reason, condition.Message)
case condition.Type == configv1.OperatorAvailable && condition.Status == configv1.ConditionFalse:
  logrus.Infof("Cluster operator {} {} is {} with {}: {}", operator.ObjectMeta.Name, condition.Type, condition.Status, condition.Reason, condition.Message)
case condition.Type == configv1.OperatorProgressing && condition.Status == configv1.ConditionTrue:
  logrus.Infof("Cluster operator {} {} is {} with {}: {}", operator.ObjectMeta.Name, condition.Type, condition.Status, condition.Reason, condition.Message)
default:
  continue
}
```

Reply:
> I feel like available==false should be an error as well. I feel like degraded but progressing should also be logged. This is only happening after waiting 30 minutes for a cluster to have formed, correct?

The current logic will log the Degraded!=False with Errorf and the Progressing!=False with Infof.

It will log Available!=True with Infof.  I [still][1] don't think that should be Errorf.  If a particular operator considers it to have been going on for long enough to be an error, they're free to set degraded to that effect.

> Is it not possible to have available=true && degraded=true?

Yup.  And install-complete [doesn't actually care about Degraded][2], although I still think we want to log it as Errorf here.

> ```
>  default:
>   continue
> ```

This approach is a whitelist of things we know we want to print.  I'd much rather have my current blacklist of things we know we don't want to print.  That way folks with surprising statuses (e.g. Available=Unknown) or types get printed Infof where we can decided whether they're noisy enough to be worth muting or not.  We're much less likely to notice and decide to unmute things we silently ignore.

[1]: https://github.com/openshift/installer/pull/2450#discussion_r332271745
[2]: https://github.com/openshift/cluster-version-operator/pull/201/files#diff-ea14ca3632b95113db2617d71fc7345dR138

Reply:
ok, I see my confusion. By your logic we only care about degraded when !available.